intra/packet_proxy.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package intra
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"net"
+	"net/netip"
+	"sync/atomic"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-go-tun2socks/intra/doh"
+	"github.com/Jigsaw-Code/outline-go-tun2socks/intra/protect"
+	"github.com/Jigsaw-Code/outline-sdk/network"
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+)
+
+type intraPacketProxy struct {
+	fakeDNSAddr netip.AddrPort
+	dns         atomic.Pointer[doh.Transport]
+	proxy       network.PacketProxy
+	listener    UDPListener
+}
+
+var _ network.PacketProxy = (*intraPacketProxy)(nil)
+
+func makeIntraPacketProxy(fakeDNS netip.AddrPort, dns doh.Transport, protector protect.Protector, listener UDPListener) (*intraPacketProxy, error) {
```

Comment:
Go usually uses "new"
```suggestion
func newIntraPacketProxy(fakeDNS netip.AddrPort, dns doh.Transport, protector protect.Protector, listener UDPListener) (*intraPacketProxy, error) {
```